### ISSUE TITLE
no-unnecesary-type-assertion: false positive for non-null-assertion and uninitialized variables

### ISSUE THREAD

**Issue by ajafff** [palantir/tslint#3521]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.1
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
// code snippet
function foo(param: string) {
  let tmp: string;
  if (param)
    tmp = param;
  tmp! === "foo"; // <- this line
}
```

with `tslint.json` configuration:

```json
"no-unnecesary-type-assertion": true
```

#### Actual behavior

Error on the non-null-assertion, because the type of the is not changed by the assertion. If you remove the assertion you get a compile error because not all code paths initialize the variable.

#### Expected behavior

No error. That's currently not possible because we need a new TypeChecker method for that: https://github.com/Microsoft/TypeScript/issues/20221

To avoid such false positives, we could exclude identifiers matching the following criteria until we get the new TypeScript API:
* assertion is on an identifier
* identifier is a mutable variable
  * let or var
  * no parameter
  * no import
  * no ambient declaration (`declare let foo`)
  * not from destructuring declaration
* declaration is in the same function scope as the assertion
* has a type annotation
* declared type does not include `| undefined`
* variable is not initialized

**Comment by ajafff** [comment]
FWIW, I have a working solution in `no-useless-assertion` of [`@fimbul/wotan`](https://www.npmjs.com/package/@fimbul/wotan) https://github.com/fimbullinter/wotan/blob/master/packages/mimir/README.md#rules

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
